src/p11_wrap.cpp

```diff
@@ -0,0 +1,276 @@
+}
+
+void P11_Session::login(const std::string &pin)
+{
+    if (isLoggedIn()) {
+        // Weird: We are trying to log in, when we are logged in already.
+        throw P11Exception(P11Exception::mismatchLoginExceptionString);
+    }
+
+    P11IsNonNegative::callChecked(lib::LibP11::P11_PKCS11_login,
+                                  _slot.get(),
+                                  0 /* Always login as normal user. */,
+                                  pin.c_str());
+}
+
+void P11_Session::logout()
+{
+    if (!isLoggedIn()) {
+        // Weird: We are trying to log out, when we are logged out already.
+        throw P11Exception(P11Exception::mismatchLoginExceptionString);
+    }
+
+    P11IsNonNegative::callChecked(lib::LibP11::P11_PKCS11_logout, _slot.get());
+}
+
+bool P11_Session::isLoggedIn()
+{
+    int result;
+    P11IsNonNegative::callChecked(
+            lib::LibP11::P11_PKCS11_is_logged_in, _slot.get(), 0 /* Check normal user. */, &result);
+
+    // Result is 1 if logged in.
+    return result == 1;
+}
+
+/***********************************************************************************
+ * Implementation of API functions.
+ */
+
+/* Destroys a PKCS11 context by first unloading it, and then freeing it. */
+static void _PKCS11_CTX_destroy(PKCS11_CTX *ctx)
+{
+    if (ctx == nullptr) {
+        throw P11Exception(P11Exception::nullCtxExceptionString);
+    }
+    lib::LibP11::P11_PKCS11_CTX_unload(ctx);
+    lib::LibP11::P11_PKCS11_CTX_free(ctx);
+}
+
+P11_PKCS11_CTX_Ptr _PKCS11_CTX_create(const std::string &module)
+{
+    PKCS11_CTX *ctx = P11CallPtr::callChecked(lib::LibP11::P11_PKCS11_CTX_new);
+    try {
+        P11IsNonNegative::callChecked(lib::LibP11::P11_PKCS11_CTX_load, ctx, module.c_str());
+    } catch (const P11Exception &e) {
+        // Free the ctx if loading failed and then throw.
+        lib::LibP11::P11_PKCS11_CTX_free(ctx);
+        throw;
+    }
+
+    // Create shared pointer which will unload and free the ctx automatically.
+    return P11_PKCS11_CTX_Ptr{ctx, _PKCS11_CTX_destroy};
+}
+
+static std::string parseID(const std::string &idHexString)
+{
+    /* We require IDs to be represented as hex strings from the user.
+     * LibP11 does the necessary pre-processing, such as unhexing, on IDs
+     * only when invoked via OpenSSL's Engine API. However, when interacting directly
+     * with LibP11 API, like we do here, this pre-processing is not performed. To make
+     * matters worse, LibP11 does not export its internal
+     * pre-processing function. We therefore begin to re-implement it here.
+     */
+    auto parsedID = boost::algorithm::unhex(idHexString);
+    return parsedID;
+}
+
+void _PKCS11_store_private_key(P11_Session &session,
+                               EVP_PKEY *pk,
+                               const std::string &label,
+                               const std::string &id)
+{
+    std::string parsedStrID = parseID(id);
+    unsigned char *parsedID =
```

Comment:
This looks a bit like this should've been put into a vector of unsigned char?

Reply:
Done.